### PR TITLE
Specify TensorBoard Data API for blob sequences

### PR DESCRIPTION
**Pull request by davidsoergel**

### CODE REVIEW

**Comment by wchargin** [comment]
@googlebot I consent.

File: tensorboard/data/provider.py [hunk 1]
```diff
@@ -331,6 +408,131 @@ def __repr__(self):
...
+class BlobSequenceDatum(object):
+  """A single datum in a blob sequence time series for a run and tag.
+
+  Attributes:
+    step: The global step at which this datum occurred; an integer. This is a
+      unique key among data of this time series.
+    wall_time: The real-world time at which this datum occurred, as `float`
+      seconds since epoch.
+    value: A tuple of blob references, providing access to elements of this
+      sequence.  The references may be either absolute URLs that can be loaded
```
[the remark this answers is not in the document]
Comment (davidsoergel) [reply tensorboard/data/provider.py]: Yeah I figured these issues would be unlikely to cause trouble in practice, but you're right, let's be more precise.

WDYT of the present solution (your first one)?

[the remark this answers is not in the document]
Comment (davidsoergel) [reply tensorboard/data/provider.py]: One open question for me: maybe we should require case-normalized keys (i.e., all lowercase), to accommodate case-insensitive filesystems.  Also, there's currently no enforcement of the spec given in the docstring; my feeling is we don't need it at this point.

File: tensorboard/data/provider.py [hunk 2]
```diff
@@ -129,12 +129,71 @@ def read_tensors(self):
     """Not yet specified."""
     pass
 
-  def list_blob_sequences(self):
-    """Not yet specified."""
+  def list_blob_sequences(self,
+                          experiment_id,
```
[the remark this answers is not in the document]
Comment (davidsoergel) [reply tensorboard/data/provider.py]: Thanks!  I bet I just introduced new formatting errors; give me a few minutes to get Black hooked up here.

File: tensorboard/data/provider.py [hunk 3]
```diff
@@ -331,6 +408,131 @@ def __repr__(self):
...
 
+class BlobSequenceTimeSeries(TimeSeries):
+  """Metadata about a blob sequence time series for a particular run and tag.
+
+  Attributes:
+    max_step: The largest step value of any datum in this scalar time series; a
+      nonnegative integer.
+    max_wall_time: The largest wall time of any datum in this time series, as
+      `float` seconds since epoch.
+    latest_max_index: The largest index in the sequence at max_step (0-based,
```
[the remark this answers is not in the document]
Comment (davidsoergel) [reply tensorboard/data/provider.py]: To be clearer: I propose to remove this field, unless there's a reason for it that I forgot/misunderstood.

File: tensorboard/data/provider.py [hunk 4]
```diff
@@ -331,6 +408,131 @@ def __repr__(self):
...
+  """A single datum in a blob sequence time series for a run and tag.
+
+  Attributes:
+    step: The global step at which this datum occurred; an integer. This is a
+      unique key among data of this time series.
+    wall_time: The real-world time at which this datum occurred, as `float`
+      seconds since epoch.
+    value: A tuple of blob references, providing access to elements of this
+      sequence.  The references may be either absolute URLs that can be loaded
+      directly, or unique keys that may later be dereferenced via
```
[the remark this answers is not in the document]
Comment (wchargin) [reply tensorboard/data/provider.py]: > even if we don't expect the caller to make use of it

Thus they’re opaque to the caller, no?

Really we’re using *opaque* here as a shorthand for an existential type,
which in my experience is fairly standard: e.g., [Flow’s opaque
types][flow], and [Standard ML’s opaque vs. transparent
ascription][sml]. (They of course can’t be opaque to _every_ layer of
the stack: something has to introspect it, even if only as a byte
sequence. They’re opaque to the caller, which is what matters.)

If you want to omit the specific phrase “opaque key”, please specify
somewhere that clients must not inspect the content of the key in any
way.

[flow]: https://flow.org/en/docs/types/opaque-types/#opaque-type-alias-syntax-
[sml]: https://www.cs.cmu.edu/~rwh/introsml/modules/sigstruct.htm

File: tensorboard/data/provider.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (wchargin) [reply tensorboard/data/provider.py]: Normalizing case helps when writing to case-insensitive filesystems, but
complicates all the other layers of the stack: you need case conversions
everywhere, and in general need to be careful that your conversions are
locale-insensitive to pass the Turkish test, and properly handle puns
like (U+00B5, U+039C, U+03BC) and (U+03A3, U+03C2, U+03C3), …. It helps
a lot that you restrict the character set to URI-valid characters;
I think that this is a good idea. But you still have poorer caching
behavior, an encoded form that’s twice as large, etc. I suspect that it
will be easier to make them case-sensitive.

File: tensorboard/data/provider.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wchargin) [reply tensorboard/data/provider.py]: Maybe we can indeed remove it entirely? I’ll think about it when I get a
chance. If you run through the request flow of the images dashboard,
does everything check out without this parameter?

File: tensorboard/data/provider.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidsoergel) [reply tensorboard/data/provider.py]: sgtm I'll leave this as a-zA-Z etc. then.

File: tensorboard/data/provider.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidsoergel) [reply tensorboard/data/provider.py]: In both of the cases you linked, the caller *cannot* know what is behind the 'opaque' curtain, which is imho the correct usage of the term.  Here, we are only *politely asking* the caller not to look (just like 'private' things in Python).

Basically I don't want to give the impression that there is any enforced secrecy here.  If somebody writes a data provider that exposes secrets in these keys, it should be clear that they made a mistake, and that this API is not somehow vouching for true opacity.

I added docstring text to this effect.

File: tensorboard/data/provider.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidsoergel) [reply tensorboard/data/provider.py]: done

File: tensorboard/data/provider.py [hunk 5]
```diff
@@ -331,6 +392,205 @@ def __repr__(self):
...
+        "max_wall_time=%r" % (self._max_wall_time,),
+        "latest_max_index=%r" % (self._latest_max_index,),
+        "plugin_content=%r" % (self._plugin_content,),
+        "description=%r" % (self._description,),
+        "display_name=%r" % (self._display_name,),
+    ))
+
+
+class BlobReference(object):
+  """A reference to a blob.
```
[the remark this answers is not in the document]
Comment (davidsoergel) [reply tensorboard/data/provider.py]: Totally agree, and fixed-- `blob_key` is required, `url` is optional.  I think we should *always* be able to serve blobs through the API, whether or not the blob is also available from a direct url.

**Pull request merged**